File: packages/agents-python/berth_agents/otel_tracer.py
"""A fourth StepTracer backend, alongside Context Bus/Semantic FS/both
(create_agent_tracer() in TypeScript's tracing.ts — not yet ported to
Python, see docs/agents-python-reference.md) — this one doesn't store
anything itself. It emits real spans through opentelemetry-api's global
tracer, so whatever OTel SDK + exporter you configure (Langfuse, Phoenix,
Honeycomb, Datadog, a plain OTel Collector, ...) receives them.
opentelemetry-api alone has no exporter and does nothing without a real SDK
registered — wiring one up is on the caller. Mirrors @berth/agents'
otel-tracer.ts field-for-field."""

from __future__ import annotations

import time
from typing import Any

from opentelemetry import trace
from opentelemetry.trace import StatusCode

from .tracing import AgentStepEvent


def _now_ns() -> int:
    return time.time_ns()


class OtelStepTracer:
    """Attribute names follow the OTel GenAI semantic conventions
    (https://opentelemetry.io/docs/specs/semconv/gen-ai/) where one
    applies — hand-written as literal strings rather than importing
    opentelemetry-semantic-conventions' GenAI constants, which are still
    marked experimental/unstable and move between module paths across SDK
    versions; the attribute name strings themselves are what's actually
    stable across a semconv version."""

    def __init__(self, tracer_name: str = "berth_agents") -> None:
        self._tracer = trace.get_tracer(tracer_name)

    async def emit(self, event: AgentStepEvent) -> None:
        end_time = _now_ns()
        start_time = end_time - event.duration_ms * 1_000_000

        if event.kind == "llm-turn":
            span_name = f"chat {event.agent_name}"
            operation = "chat"
        else:
            span_name = f"execute_tool {event.tool_name or 'unknown'}"
            operation = "execute_tool"

        attributes: dict[str, Any] = {
            "gen_ai.operation.name": operation,
            "gen_ai.agent.name": event.agent_name,
            # Not part of any OTel semantic convention — Berth-specific, and
            # the one thing that lets a backend correlate every span from
            # the same Agent.run()/Crew composition. There's no single
            # parent span linking them: emit() has no signal for when a run
            # begins or ends to build one.
            "berth.run_id": event.run_id,
            "berth.turn": event.turn,
        }
        if event.tool_name:
            attributes["gen_ai.tool.name"] = event.tool_name
        if event.usage:
            attributes["gen_ai.usage.input_tokens"] = event.usage.input_tokens
            attributes["gen_ai.usage.output_tokens"] = event.usage.output_tokens

        span = self._tracer.start_span(span_name, start_time=start_time, attributes=attributes)
        if event.error:
            span.record_exception(Exception(event.error), timestamp=end_time)
            span.set_status(StatusCode.ERROR, event.error)
        else:
            span.set_status(StatusCode.OK)
        span.end(end_time)


def create_otel_step_tracer(*, tracer_name: str = "berth_agents") -> OtelStepTracer:
    return OtelStepTracer(tracer_name=tracer_name)

File: packages/agents/src/otel-tracer.ts
import { trace, SpanStatusCode, type Attributes } from "@opentelemetry/api";
import type { AgentStepEvent, StepTracer } from "./tracing.js";

export interface OtelStepTracerOptions {
  /** Registered as the instrumentation scope name with the global TracerProvider — cosmetic, shows up in most backends' UI as "where this span came from." */
  tracerName?: string;
}

/**
 * A fourth `StepTracer` backend, alongside Context Bus/Semantic FS/both
 * (`createAgentTracer()`) — this one doesn't store anything itself. It emits
 * real spans through `@opentelemetry/api`'s global tracer, so whatever OTel
 * SDK + exporter *you* configure (Langfuse, Phoenix, Honeycomb, Datadog, a
 * plain OTel Collector, ...) receives them. `@opentelemetry/api` alone has
 * no exporter and does nothing without a real SDK registered — same
 * "instrumentation library, not a vendored backend" posture every other
 * OTel-emitting library has; wiring up `@opentelemetry/sdk-node` (or
 * whatever your own app already uses) is on the caller, not this package.
 *
 * Attribute names follow the OTel GenAI semantic conventions
 * (https://opentelemetry.io/docs/specs/semconv/gen-ai/) where one applies —
 * hand-written as literal strings rather than importing
 * `@opentelemetry/semantic-conventions`'s GenAI constants, which are still
 * marked experimental/unstable and move between package paths across SDK
 * versions; the attribute name strings themselves are what's actually
 * stable across a semconv version, so this is the safer dependency to take.
 */
export function createOtelStepTracer(options: OtelStepTracerOptions = {}): StepTracer {
  const tracer = trace.getTracer(options.tracerName ?? "@berth/agents");

  return {
    async emit(event: AgentStepEvent): Promise<void> {
      const endTime = new Date();
      const startTime = new Date(endTime.getTime() - event.durationMs);
      // The two-way ternary this replaces labelled anything that wasn't an
      // llm-turn as `execute_tool unknown`, so the compaction events
      // REMEDIATION 4.1 added would have arrived in every backend as
      // phantom tool calls to a tool named "unknown".
      const spanName =
        event.kind === "llm-turn"
          ? `chat ${event.agentName}`
          : event.kind === "tool-call"
            ? `execute_tool ${event.toolName ?? "unknown"}`
            : `compact_context ${event.agentName}`;

      const attributes: Attributes = {
        "gen_ai.agent.name": event.agentName,
        // Not part of any OTel semantic convention — Berth-specific, and
        // the one thing that lets a backend correlate every span from the
        // same Agent.run()/Crew composition. See the module doc comment in
        // tracing.ts for why there's no single parent span linking them:
        // emit() has no signal for when a run begins or ends to build one.
        "berth.run_id": event.runId,
        "berth.turn": event.turn,
      };
      // Only the two operations the GenAI semantic conventions actually
      // define get a `gen_ai.operation.name`. Compaction has no equivalent
      // there, and inventing a value would put a non-standard string in the
      // one attribute backends switch on to classify a span.
      if (event.kind === "llm-turn") attributes["gen_ai.operation.name"] = "chat";
      if (event.kind === "tool-call") attributes["gen_ai.operation.name"] = "execute_tool";
      if (event.droppedMessages !== undefined) attributes["berth.dropped_messages"] = event.droppedMessages;
      if (event.toolName) attributes["gen_ai.tool.name"] = event.toolName;
      if (event.usage) {
        attributes["gen_ai.usage.input_tokens"] = event.usage.inputTokens;
        attributes["gen_ai.usage.output_tokens"] = event.usage.outputTokens;
      }

      const span = tracer.startSpan(spanName, { startTime, attributes });
      if (event.error) {
        span.recordException(event.error, endTime);
        span.setStatus({ code: SpanStatusCode.ERROR, message: event.error });
      } else {
        span.setStatus({ code: SpanStatusCode.OK });
      }
      span.end(endTime);
    },
  };
}
